src/core/core/flags.cpp

```diff
@@ -0,0 +1,130 @@
+    FV_INT32 = 1,
+    FV_UINT32 = 2,
+    FV_INT64 = 3,
+    FV_UINT64 = 4,
+    FV_DOUBLE = 5,
+    FV_STRING = 6,
+    FV_MAX_INDEX = 6,
+};
+
+using validator_fn = std::function<void()>;
+
+class flag_data
+{
+public:
+#define FLAG_DATA_LOAD_CASE(type, type_enum, suffix)                                               \
+    case type_enum:                                                                                \
+        value<type>() = dsn_config_get_value_##suffix(_section, _name, value<type>(), _desc);      \
+        if (_validator) {                                                                          \
+            _validator();                                                                          \
+        }                                                                                          \
+        break
+
+    void load()
+    {
+        switch (_type) {
+            FLAG_DATA_LOAD_CASE(int32_t, FV_INT32, int64);
+            FLAG_DATA_LOAD_CASE(int64_t, FV_INT64, int64);
+            FLAG_DATA_LOAD_CASE(uint32_t, FV_UINT32, uint64);
+            FLAG_DATA_LOAD_CASE(uint64_t, FV_UINT64, uint64);
+            FLAG_DATA_LOAD_CASE(bool, FV_BOOL, bool);
+            FLAG_DATA_LOAD_CASE(double, FV_DOUBLE, double);
+            FLAG_DATA_LOAD_CASE(const char *, FV_STRING, string);
+        }
+    }
+
+    flag_data(const char *section, const char *name, const char *desc, value_type type, void *val)
+        : _type(type), _val(val), _section(section), _name(name), _desc(desc)
+    {
+    }
+
+    void set_validator(validator_fn &validator) { _validator = std::move(validator); }
+    const validator_fn &validator() const { return _validator; }
+
+private:
+    template <typename T>
+    T &value()
+    {
+        return *reinterpret_cast<T *>(_val);
+    }
+
+private:
+    const value_type _type;
+    void *const _val;
+    const char *_section;
+    const char *_name;
+    const char *_desc;
+    validator_fn _validator;
+};
+
+class flag_registry : public utils::singleton<flag_registry>
+{
+public:
+    void add_flag(const char *name, flag_data flag) { _flags.emplace(name, flag); }
+
+    void add_validator(const char *name, validator_fn &validator)
+    {
+        auto it = _flags.find(name);
+        dassert(it != _flags.end(), "flag \"%s\" does not exist", name);
+        flag_data &flag = it->second;
+        if (!flag.validator()) {
+            flag.set_validator(validator);
+        }
+    }
+
+    void load_from_config()
+    {
+        for (auto &kv : _flags) {
+            flag_data &flag = kv.second;
+            flag.load();
+        }
+    }
+
+private:
+    friend class utils::singleton<flag_registry>;
+    flag_registry() = default;
+
+private:
+    std::map<std::string, flag_data> _flags;
+};
+
+#define FLAG_REG_CONSTRUCTOR(type, type_enum)                                                      \
+    flag_registerer::flag_registerer(                                                              \
```

Comment:
why the name is `flag_registerer`, but not `flag_register`

Reply:
These class names are learned from gflags. If users are familiar with gflags, he will soon get started to use dsn flags.

See this: https://github.com/gflags/gflags/blob/master/src/gflags.h.in#L432.

"registerer" is the role to register the flag, where "register" is a verb, not a noun.